lib/block_view.go

```diff
@@ -4728,133 +5592,789 @@ func (bav *UtxoView) _connectUpdateProfile(
+	if bidderChangeNanos > 0 {
+		nextUtxoIndex += 1
+		bidderChangeOutputKey := &UtxoKey{
+			TxID:  *txHash,
+			Index: nextUtxoIndex,
+		}
+
+		utxoEntry := UtxoEntry{
+			AmountNanos: bidderChangeNanos,
+			PublicKey:   bidderPublicKey,
+			BlockHeight: blockHeight,
+			UtxoType:    UtxoTypeNFTCreatorRoyalty,
+
+			UtxoKey: bidderChangeOutputKey,
+			// We leave the position unset and isSpent to false by default.
+			// The position will be set in the call to _addUtxo.
+		}
+
+		_, err = bav._addUtxo(&utxoEntry)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(err, "_connectAcceptNFTBid: Problem adding output utxo")
 		}
+		nftPaymentUtxoKeys = append(nftPaymentUtxoKeys, bidderChangeOutputKey)
 	}
-	// Save a copy of the profile entry so so that we can safely modify it.
-	var prevProfileEntry *ProfileEntry
-	if existingProfileEntry != nil {
-		// NOTE: The only pointer in here is the StakeEntry and CoinEntry pointer, but since
-		// this is not modified below we don't need to make a copy of it.
-		prevProfileEntry = &ProfileEntry{}
-		*prevProfileEntry = *existingProfileEntry
+
+	// (6) Add creator coin royalties to bitclout locked. If the number of coins in circulation is
+	// less than the "auto sell threshold" we burn the bitclout.
+	if creatorCoinRoyaltyNanos > 0 && existingProfileEntry.CoinsInCirculationNanos >= bav.Params.CreatorCoinAutoSellThresholdNanos {
+		// Make a copy of the previous coin entry. It has no pointers, so a direct copy is ok.
+		newCoinEntry := prevCoinEntry
+		newCoinEntry.BitCloutLockedNanos += creatorCoinRoyaltyNanos
+		existingProfileEntry.CoinEntry = newCoinEntry
+		bav._setProfileEntryMappings(existingProfileEntry)
 	}
 
-	// If a profile already exists then we only update fields that are set.
-	var newProfileEntry ProfileEntry
-	if existingProfileEntry != nil && !existingProfileEntry.isDeleted {
-		newProfileEntry = *existingProfileEntry
+	// (7) Save a copy of the previous postEntry and then decrement NumNFTCopiesForSale.
+	prevPostEntry := &PostEntry{}
+	*prevPostEntry = *nftPostEntry
+	nftPostEntry.NumNFTCopiesForSale--
+	bav._setPostEntryMappings(nftPostEntry)
 
-		// Modifying a profile is only allowed if the transaction public key equals
-		// the profile public key or if the public key belongs to a paramUpdater.
-		_, updaterIsParamUpdater := bav.Params.ParamUpdaterPublicKeys[MakePkMapKey(txn.PublicKey)]
-		if !reflect.DeepEqual(txn.PublicKey, existingProfileEntry.PublicKey) &&
-			!updaterIsParamUpdater {
+	// Add an operation to the list at the end indicating we've connected an NFT bid.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:                      OperationTypeAcceptNFTBid,
+		PrevNFTEntry:              prevNFTEntry,
+		PrevPostEntry:             prevPostEntry,
+		PrevCoinEntry:             &prevCoinEntry,
+		DeletedNFTBidEntries:      deletedBidEntries,
+		NFTPaymentUtxoKeys:        nftPaymentUtxoKeys,
+		PrevAcceptedNFTBidEntries: prevAcceptedBidHistory,
+	})
```

Comment:
I would add a *hardcore* sanity-check here that verifies that no free money was printed. It should work as follows:

1) Compute:
  - sellerBalanceAfter - sellerBalanceBefore
  - bidderBalanceAfter - bidderBalanceBefore (only relevant if bidder != seller)
  - creatorBalanceAfter - creatorBalanceBefore (only relevant if creator != seller and creator != bidder)
  - newCoinEntry.BitCloutLockedNanos - prevCoinEntry.BitCloutLockedNanos
 2) Verify that the sum of all of the above is <= 0 (or something like that).